### PR TITLE
More robust logging in case of corrupt CM

### PR DESCRIPTION
**Pull request by stanlyDoge**
### Type of change

- Enhancement / new feature

### Description
This PR provides check for 
- empty values
- char values instead of integer
- typos in enumerated values

I need to do some tests, so this can be marked as WIP.
### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging

### CODE REVIEW

File: cluster-controller/src/main/java/io/strimzi/controller/cluster/model/AbstractModel.java [hunk 1]
```diff
@@ -626,4 +627,39 @@ protected void kafkaHeapOptions(List<EnvVar> envVars, double dynamicHeapFraction
...
+        } catch (NumberFormatException e)  {
+            String msg = " is corrupted.";
+            if (data.get(field).toString().length() == 0) {
+                msg = " is empty.";
+            }
+            throw new InvalidConfigMapException(field + msg);
+        }
+    }
+
+    public static void checkType(Map<String, Object> data, String field) {
```
Comment (tombentley) [inline cluster-controller/src/main/java/io/strimzi/controller/cluster/model/AbstractModel.java]: I don't understand the name `checkType`, surely it should be `checkStorage`? And of course the same comment about validate+get applies.

File: cluster-controller/src/main/java/io/strimzi/controller/cluster/model/KafkaCluster.java [hunk 2]
```diff
@@ -121,12 +156,13 @@ public static KafkaCluster fromConfigMap(ConfigMap kafkaClusterCm) {
                 Labels.fromResource(kafkaClusterCm));
 
         Map<String, String> data = kafkaClusterCm.getData();
+        checkAll(data);
         kafka.setReplicas(Integer.parseInt(data.getOrDefault(KEY_REPLICAS, String.valueOf(DEFAULT_REPLICAS))));
```
Comment (tombentley) [inline cluster-controller/src/main/java/io/strimzi/controller/cluster/model/KafkaCluster.java]: So here at the call site we'd have

```
kafka.setReplicas(getInteger(data, KEY_REPLICAS, DEFAULT_REPLICAS));
```

Which is shorter, more readable, and avoids the slightly crazy convert-`DEFAULT_REPLICAS`-to-a-string-so-we-can-call-`parseInteger`-with-it.

File: cluster-controller/src/main/java/io/strimzi/controller/cluster/model/ZookeeperCluster.java [hunk 3]
```diff
@@ -136,6 +137,24 @@ public static ZookeeperCluster fromConfigMap(ConfigMap kafkaClusterCm) {
...
+        checkSingleNumber(data, KEY_REPLICAS);
+        checkSingleNumber(data, KEY_HEALTHCHECK_DELAY);
+        checkSingleNumber(data, KEY_HEALTHCHECK_TIMEOUT);
+
+        checkType(data, KEY_STORAGE);
+
+        checkString(data, KEY_IMAGE);
+        checkString(data, KEY_METRICS_CONFIG);
+
+
```
Comment (tombentley) [inline cluster-controller/src/main/java/io/strimzi/controller/cluster/model/ZookeeperCluster.java]: Needless blank line

File: cluster-controller/src/main/java/io/strimzi/controller/cluster/operator/assembly/AbstractAssemblyOperator.java [hunk 4]
```diff
@@ -43,7 +43,6 @@
     protected final boolean isOpenShift;
     protected final AssemblyType assemblyType;
     protected final ConfigMapOperator configMapOperations;
-
```
Comment (tombentley) [inline cluster-controller/src/main/java/io/strimzi/controller/cluster/operator/assembly/AbstractAssemblyOperator.java]: Please try to avoid unnecessary whitespace changes (I'm probably guilty of this at times too)

File: cluster-controller/src/test/java/io/strimzi/controller/cluster/model/KafkaConnectClusterTest.java [hunk 5]
```diff
@@ -181,4 +181,53 @@ public void testGenerateDeployment()   {
...
+
+    @Test
+    public void testCorruptedValues() {
+        ConfigMap cm = ResourceUtils.createEmptyKafkaConnectClusterConfigMap(namespace, cluster);
+
+        // type mismatch
+        cm.getData().put("kafka-healthcheck-delay", "1z");
+        try {
+            KafkaCluster kc = KafkaCluster.fromConfigMap(cm);
+        } catch (Throwable t) {
```
Comment (tombentley) [inline cluster-controller/src/test/java/io/strimzi/controller/cluster/model/KafkaConnectClusterTest.java]: This doesn't actually test that `fromConfigMap()` is throwing anything – if it doesn't throw the `catch` will be skipped and the test will still pass. Also you should only catch the specific exception that you want to assert. So you need to write it like this:

```
try {
    ...
    fail("Expected it to throw");
} catch (InvalidConfigMapException e) {
    // any assertions on message
}
```

The other way you can do it is to write `@Test(expect=InvalidConfigMapException.class)`-annotated methods for each of the fields. You end up with more tests that way (so it's more verbose), but each of them are more specific. 

Generally speaking it's a bad idea to assert equality on an exception message and it makes changing the exception message very painful is it will break all the tests. In this case `InvalidConfigMapException` should have a property for the field as well as the message. Then you can assert that the exception is about the expected key, without caring how the message is worded. It will also fit with the events stuff when we get to it.

File: cluster-controller/src/test/java/io/strimzi/controller/cluster/model/KafkaConnectClusterTest.java [hunk 6]
```diff
@@ -181,4 +181,53 @@ public void testGenerateDeployment()   {
...
+            assertEquals("num.io.threads is corrupted.", t.getMessage());
+        }
+
+        // corrupted JSON (missing quotes)
+        cm.getData().clear();
+        cm.getData().put("kafka-config", "{  \"num.io.threads\": lol }");
+        try {
+            KafkaCluster kc = KafkaCluster.fromConfigMap(cm);
+        } catch (Throwable t) {
+            assertEquals("Config Map JSON is corrupted.", t.getMessage());
```
Comment (tombentley) [inline cluster-controller/src/test/java/io/strimzi/controller/cluster/model/KafkaConnectClusterTest.java]: Why don't we include the "kafka-config" key name here?

File: cluster-controller/src/test/java/io/strimzi/controller/cluster/operator/assembly/KafkaConnectAssemblyOperatorTest.java [hunk 7]
```diff
@@ -405,7 +405,6 @@ public void testReconcile(TestContext context) {
         ServiceOperator mockServiceOps = mock(ServiceOperator.class);
         DeploymentOperator mockDcOps = mock(DeploymentOperator.class);
 
-
```
Comment (tombentley) [inline cluster-controller/src/test/java/io/strimzi/controller/cluster/operator/assembly/KafkaConnectAssemblyOperatorTest.java]: pointless ws change

File: cluster-controller/src/test/java/io/strimzi/controller/cluster/operator/assembly/KafkaConnectS2IAssemblyOperatorTest.java [hunk 8]
```diff
@@ -564,7 +564,6 @@ public void testReconcile(TestContext context) {
         BuildConfigOperator mockBcOps = mock(BuildConfigOperator.class);
         ImageStreamOperator mockIsOps = mock(ImageStreamOperator.class);
 
-
```
Comment (tombentley) [inline cluster-controller/src/test/java/io/strimzi/controller/cluster/operator/assembly/KafkaConnectS2IAssemblyOperatorTest.java]: pointless ws change

**Comment by stanlyDoge** [comment]
Checkstyle limits were temporarily bumped. This will be reverted after AbstractModel refactoring.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperatorTest.java [hunk 9]
```diff
@@ -564,9 +564,9 @@ public void testReconcile(TestContext context) {
         BuildConfigOperator mockBcOps = mock(BuildConfigOperator.class);
         ImageStreamOperator mockIsOps = mock(ImageStreamOperator.class);
 
-
         String clusterCmNamespace = "test";
 
+
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectS2IAssemblyOperatorTest.java]: Why add extra line here?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperatorTest.java [hunk 10]
```diff
@@ -405,9 +405,9 @@ public void testReconcile(TestContext context) {
         ServiceOperator mockServiceOps = mock(ServiceOperator.class);
         DeploymentOperator mockDcOps = mock(DeploymentOperator.class);
 
-
         String clusterCmNamespace = "test";
 
+
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperatorTest.java]: Why add extra line here?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 11]
```diff
@@ -626,4 +628,84 @@ protected void kafkaHeapOptions(List<EnvVar> envVars, double dynamicHeapFraction
...
+            }
+        }
+        return blame;
+    }
+
+    public static String getConfig(Map<String, String> data, String key) {
+        String config = "";
+        try {
+            config = data.get(key);
+            Map<String, Object> map = new JsonObject(config).getMap();
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: So that you know: The CO started off using the Vertx JSON parsing API, but we're now drifting in the direction of using the Jackson API directly. The Jackson API exposes a little more information in its exceptions, which might have made `getJSONcorruptionBlame()` easier to write and less fragile. Generally speaking it's not good practice to try to parse exception messages for the same reason that it's not great to assert on them: It will break should the format of the exception message change. In this case it's especially bad because the messages are coming from a dependency you don't control so any breakage is outside your control.

I think it's OK to leave this using the Vertx Json API for now, but it's something we will have to tidy up, so please could you open an issue for it.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 12]
```diff
@@ -118,26 +122,23 @@ public static KafkaCluster fromConfigMap(ConfigMap kafkaClusterCm) {
...
-        String storageConfig = data.get(KEY_STORAGE);
-        kafka.setStorage(Storage.fromJson(new JsonObject(storageConfig)));
+        kafka.setStorage(getStorage(data, KEY_STORAGE));
 
-        String kafkaConfig = data.get(KEY_KAFKA_CONFIG);
-        if (kafkaConfig != null) {
-            kafka.setConfiguration(new KafkaConfiguration(new JsonObject(kafkaConfig)));
-        }
+        String kafkaConfig = getConfig(data, KEY_KAFKA_CONFIG);
+        kafka.setConfiguration(new KafkaConfiguration(new JsonObject(kafkaConfig)));
```
[the remark this answers is not in the document]
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: Actually, the `Properties` class is kind of messy. I think it was not the best choice to use it in `AbstractConfiguration`. While it makes some things easier, it also brings a lot of limitations because it is opinionated and expects the use of the properties file only with Java. I'm afraid that one day we will need to get rid of ti again. We should not spread the use of Properties outside of `Configuration` classes into the `Model` classes.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/InvalidConfigMapException.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+ * Copyright 2017-2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package io.strimzi.operator.cluster;
+
+public class InvalidConfigMapException extends RuntimeException {
+
+    private String key;
+    public InvalidConfigMapException(String message, String key) {
```
[the remark this answers is not in the document]
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/InvalidConfigMapException.java]: I had this idea too, but i was lazy do that :(

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: You're right that `Properties` shows its age and isn't fun to use. But currently the `AbstractConfiguration`'s subclasses are all Java things which take `.properties` files for their configuration: I don't think there's much point in trying to encapsulate that. When we have some component which consumes its config as YAML or XML or whatever else that would either be a different thing than an `AbstractConfiguration`, or we would at that point generify `AbstractConfiguration` to not care about the type.

I think it's perfectly reasonable to validate that the value in the CM conforms to something we can convert to a `.properties` file, and it's natural to do that by returning a `Properties` instance at the same time. The fact that `Properties` is old and ugly doesn't negate that.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/InvalidConfigMapException.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/InvalidConfigMapException.java]: In intellij: Right click > Refactor > Change signature. Or the equivalent in Eclipse.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: It is not about being old and ugly. The `Properties` class is designed primarily to write and load the properties from a single application. Not as a mean to create properties files. As a result it does things like comments, escaping and has a bad handling of default values.

AFAIK using `Properties` outside of the `Configuration` classes will anyway not give us any benefits compared to `Map<String, String>` which it wraps.

[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: > Not as a mean to create properties files.

I don't understand what you mean here. It *defines* the format for a `.properties` file.

> compared to `Map<String, String>` which it wraps.

I could live with that, I suppose.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 14]
```diff
@@ -626,4 +628,86 @@ protected void kafkaHeapOptions(List<EnvVar> envVars, double dynamicHeapFraction
...
+        for (i = 0; i < entries.length; ++i) {
+            if (entries[i].split(":")[1].equals(token)) {
+                blame = entries[i].split(":")[0];
+                break;
+            }
+        }
+        return blame;
+    }
+
+    public static JsonObject getConfig(Map<String, String> data, String key) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: Does this method work then the `key` doesn't exist? In that case the try block would be probably throwing some NPEs or something similar.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 15]
```diff
@@ -626,4 +628,86 @@ protected void kafkaHeapOptions(List<EnvVar> envVars, double dynamicHeapFraction
...
+        String config = "";
+        try {
+            config = data.get(key);
+            Map<String, Object> map = new JsonObject(config).getMap();
+            Map<String, String> newMap = new HashMap<String, String>();
+            for (Map.Entry<String, Object> entry : map.entrySet()) {
+                if (entry.getValue() instanceof String) {
+                    newMap.put(entry.getKey(), (String) entry.getValue());
+                } else {
+                    throw new InvalidConfigMapException(entry.getKey(), " has JSON with key " + entry.getKey() + " with a non-String value");
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: The JSON string with the config expects to receive also booleans, floats etc. These should be properly converted.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 16]
```diff
@@ -118,26 +118,23 @@ public static KafkaCluster fromConfigMap(ConfigMap kafkaClusterCm) {
...
-        String storageConfig = data.get(KEY_STORAGE);
-        kafka.setStorage(Storage.fromJson(new JsonObject(storageConfig)));
+        kafka.setStorage(getStorage(data, KEY_STORAGE));
 
-        String kafkaConfig = data.get(KEY_KAFKA_CONFIG);
-        if (kafkaConfig != null) {
-            kafka.setConfiguration(new KafkaConfiguration(new JsonObject(kafkaConfig)));
-        }
+        JsonObject kafkaConfig = getConfig(data, KEY_KAFKA_CONFIG);
+        kafka.setConfiguration(new KafkaConfiguration(kafkaConfig));
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: If we are changing the configuration from using `null` to being empty but not null, we should maybe remove the unnecessary `null` check here: https://github.com/strimzi/strimzi/blob/23a12d8452666963ab5cbdedd2fb2b5102080670/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java#L304

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+            Map<String, String> stringMap = new HashMap<String, String>();
+            for (Map.Entry<String, Object> entry : map.entrySet()) {
+                if (entry.getValue() instanceof String) {
+                    stringMap.put(entry.getKey(), (String) entry.getValue());
+                } else if (entry.getValue() instanceof Integer || entry.getValue() instanceof Long || entry.getValue() instanceof Double || entry.getValue() instanceof Float) {
+                    numMap.put(entry.getKey(), String.valueOf(entry.getValue()));
+                } else if (entry.getValue() instanceof Boolean) {
+                    boolMap.put(entry.getKey(), String.valueOf(entry.getValue()));
+                }  else {
+                    throw new InvalidConfigMapException(entry.getKey(), " has JSON with key " + entry.getKey() + " with a non-String value");
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: We now allow even other values than just String to pass, so we should fix the error message in the Exception (i.e. the `non-string value` doesn't apply anymore).

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+                }
+            }
+
+            for (Map.Entry<String, String> entry : numMap.entrySet()) {
+                getInteger(numMap, entry.getKey(), 0);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: In the loop above which validates the Objects we pass `Integer`, `Long`, `Double` and `Float`. But here we throw an exception when it is anything else than `Integer`. These should be in sync.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    public static JsonObject getConfig(Map<String, String> data, String key) {
+        String config = data.get(key);
+        if (config == null) {
+            return null;
+        }
+        try {
+            Map<String, Object> map = new JsonObject(config).getMap();
+            Map<String, String> numMap = new HashMap<String, String>();
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: I do not understand the purpose of the `numMap`, `boolMap` and `stringMap` and the secondary validation using the `getInteger` etc.methods. The `JsonObject` already validated these. If the `JsonObject` class is able to convert the value to `Integer` or `Boolean` it has to be `Integer` or `Boolean`. 

Why do we have to double check it afterwards? Can't we just remove the lines 109-120?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 20]
```diff
@@ -118,25 +118,24 @@ public static KafkaCluster fromConfigMap(ConfigMap kafkaClusterCm) {
...
-        kafka.setHealthCheckTimeout(Integer.parseInt(data.getOrDefault(KEY_HEALTHCHECK_TIMEOUT, String.valueOf(DEFAULT_HEALTHCHECK_TIMEOUT))));
+        kafka.setReplicas(Utils.getInteger(data, KEY_REPLICAS, DEFAULT_REPLICAS));
+        kafka.setImage(Utils.getNonemptyString(data, KEY_IMAGE, DEFAULT_IMAGE));
+        kafka.setHealthCheckInitialDelay(Utils.getInteger(data, KEY_HEALTHCHECK_DELAY, DEFAULT_HEALTHCHECK_DELAY));
+        kafka.setHealthCheckTimeout(Utils.getInteger(data, KEY_HEALTHCHECK_TIMEOUT, DEFAULT_HEALTHCHECK_TIMEOUT));
 
-        kafka.setZookeeperConnect(data.getOrDefault(ENV_VAR_KAFKA_ZOOKEEPER_CONNECT, kafkaClusterCm.getMetadata().getName() + "-zookeeper:2181"));
+        kafka.setZookeeperConnect(kafkaClusterCm.getMetadata().getName() + "-zookeeper:2181");
 
         String metricsConfig = data.get(KEY_METRICS_CONFIG);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: We seem to use `Utils.getConfig` for metrics in `ZookeeperCluster`. It should do the same on both places.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        for (i = 0; i < entries.length; ++i) {
+            if (entries[i].split(":")[1].equals(token)) {
+                blame = entries[i].split(":")[0];
+                break;
+            }
+        }
+        return blame;
+    }
+
+    public static JsonObject getConfig(Map<String, String> data, String key) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: Leaving other comments aside ... I have major doubts about this method. It will not work as designed now. You are trying to validate too much in the same method. For example:

The `kafka-config`, `connect-config` and `zookeeper-config` fields have completely different validation requirements than the `kafka-metrics-config` and `zookeeper-metrics-config` fields.

The config fields are just key-value maps. But the metrics are a real JSON objects. So for example the metrics config cannot be validated by this method because it can contain another JSON objects or arrays. When validating the regular configs, this method can work. But it basically duplicates the functionality in `AbstractConfiguration` with the only difference that it throws exceptions instead of ignoring the invalid fields with error message.

My two cents are that we should:
* Have a generic method which validates that the field is JsonObject
* Change the `AbstractConfiguration` to throw exceptions for invalid fields inside the JSON and use it in a method similar to `Utils.getStorage`

PS: Have you tried to run this with real cluster? I haven't but my guess is that it would fail with the example YAML files.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java (hunk 19 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: At first, i thouhgt there are just number values in whatever-config, so it was designed to iterate all values and check if its number. Yes, we can delete this lines, but issue is, if I put value "string123" to some number key, JsonObject parses it ok.

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: You mean if someone does something like this?
```json
{
"this-shoutld-be-a-number": "string123"
}
```

But in most cases you anyway don't know whether the key should be a number or string. So you cannot validate this on this level. (we actually never validate this for the configuration or metrics fields).

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java (hunk 20 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: Yes, i missed this one :(

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+        for (i = 0; i < entries.length; ++i) {
+            if (entries[i].split(":")[1].equals(token)) {
+                blame = entries[i].split(":")[0];
+                break;
+            }
+        }
+        return blame;
+    }
+
+    public static <T extends AbstractConfiguration> T getConfig(Map<String, String> data, String key) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: We should use either "config" or "configuration" consistently. 

And TBH I don't think this method achieves much: I think it would be better either with separate `getKafkaConfuration()`, `getZookeeperConfuration()` and `getKafkaConnectConfuration()` methods, avoiding the `if/else if`. Or, alternatively, you could just have the current callers of this method call `getJson()` directly and instantiate the relevant type of `*Configuration` themselves.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+
+    public static JsonObject getJson(Map<String, String> data, String key) {
+        String config = data.get(key);
+        if (config == null) {
+            return null;
+        }
+        try {
+            new JsonObject(config);
+        } catch (DecodeException de) {
+            throw new InvalidConfigMapException(getJsonCorruptionBlame(de, config), " corruptes JSON");
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: Should it be `corrupted JSON`? (typo)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Storage.java [hunk 24]
```diff
@@ -177,7 +177,7 @@ public static StorageType from(String type) {
             } else if (type.equals(LOCAL.type)) {
                 return LOCAL;
             } else {
-                throw new IllegalArgumentException("Unknown type: " + type);
+                throw new IllegalArgumentException("Unknown type: " + type + ". Allowed types are: " + EPHEMERAL + ", " + PERSISTENT_CLAIM + " and " + LOCAL + ".");
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Storage.java]: Maybe we can use the implicit `values()` method in the `StorageType` enum for printing the available types ?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+            if (data.get(key) == null) { // value is not set in config map
+                return new Storage(Storage.StorageType.PERSISTENT_CLAIM);
+            }
+            return Storage.fromJson(new JsonObject(data.get(key)));
+        } catch (Exception e) {
+            throw new InvalidConfigMapException(key, " is corrupted");
+        }
+    }
+
+    public static String getNonemptyString(Map<String, String> data, String key, String defaultValue) {
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: more readable `getNonEmptyString` than `getNonemptyString` ?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java (hunk 23 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: Lets say ```key.name``` has invalid syntax. It breaks JSON and it says ```key.name corruptes JSON```. So it is ok, I guess.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Storage.java (hunk 24 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Storage.java]: Sure. Thanks!

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java (hunk 25 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: Fixed.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java (hunk 23 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Utils.java]: Ah, ok ... isn't it `corrupts` in that case? I don't know ... _corruptes_ sounds terribly German :-).
